diff --git a/src/loadlib.c b/src/loadlib.c
index 6d289fc..0e43689 100644
--- a/src/loadlib.c
+++ b/src/loadlib.c
@@ -122,7 +122,7 @@ static void lsys_unloadlib (void *lib) {
 
 
 static void *lsys_load (lua_State *L, const char *path, int seeglb) {
-  void *lib = dlopen(path, RTLD_NOW | (seeglb ? RTLD_GLOBAL : RTLD_LOCAL));
+  void *lib = dlopen(*path? path : NULL, RTLD_NOW | (seeglb ? RTLD_GLOBAL : RTLD_LOCAL));
   if (l_unlikely(lib == NULL))
     lua_pushstring(L, dlerror());
   return lib;
@@ -198,7 +198,7 @@ static void lsys_unloadlib (void *lib) {
 
 
 static void *lsys_load (lua_State *L, const char *path, int seeglb) {
-  HMODULE lib = LoadLibraryExA(path, NULL, LUA_LLE_FLAGS);
+  HMODULE lib = *path? LoadLibraryExA(path, NULL, LUA_LLE_FLAGS) : GetModuleHandle(NULL);
   (void)(seeglb);  /* not used: symbols are 'global' by default */
   if (lib == NULL) pusherror(L);
   return lib;
